components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,297 @@
+
+/**
+ * Gets the size of a file specified by `file_path` and adds it to file section `offset`.
+ * @param file_path
+ * @param[out] offset File section offset for the single-file archive. The returned offset
+ * represents the starting position of the next file in single-file archive.
+ * @throws OperationFailed if error getting file size.
+ */
+auto get_file_size_and_update_offset(std::filesystem::path const& file_path, uint64_t& offset)
+        -> void;
+
+/**
+ * Generates metadata for the file section of a single-file archive. The metadata consists
+ * of a list of file names and their corresponding starting offsets.
+ *
+ * @param multi_file_archive_path
+ * @param segment_ids
+ * @return Vector containing a `FileInfo` struct for every file in the multi-file archive.
+ * @throws Propagates `update_offset`'s exceptions.
+ */
+[[nodiscard]] auto get_file_infos(
+        std::filesystem::path const& multi_file_archive_path,
+        std::vector<std::string> const& segment_ids
+) -> std::vector<FileInfo>;
+
+/**
+ * Combines file section metadata, multi-file archive metadata, and the number of segments into
+ * single-file archive metadata. Once combined, serializes the metadata into MsgPack format.
+ *
+ * @param multi_file_archive_metadata
+ * @param file_infos Vector containing a `FileInfo` struct for every file in the multi-file archive.
+ * @param num_segments
+ * @return Packed metadata.
+ */
+[[nodiscard]] auto pack_single_file_archive_metadata(
+        ArchiveMetadata const& multi_file_archive_metadata,
+        std::vector<FileInfo> const& file_infos,
+        size_t num_segments
+) -> std::stringstream;
+
+/**
+ * Writes single-file archive header.
+ *
+ * @param archive_writer
+ * @param packed_metadata_size
+ */
+auto write_archive_header(FileWriter& archive_writer, size_t packed_metadata_size) -> void;
+
+/**
+ * Writes single-file archive metadata.
+ *
+ * @param archive_writer
+ * @param packed_metadata Packed metadata.
+ */
+auto write_archive_metadata(FileWriter& archive_writer, std::stringstream const& packed_metadata)
+        -> void;
+
+/**
+ * Reads the content of a file and writes it to the single-file archive.
+ * @param file_path
+ * @param archive_writer
+ * @throws OperationFailed if reading the file fails.
+ */
+auto write_archive_file(std::string const& file_path, FileWriter& archive_writer) -> void;
+
+/**
+ * Iterates over files in the multi-file archive copying their contents to the single-file archive.
+ * Skips metadata file since already written in `write_archive_metadata`.
+ *
+ * @param archive_writer
+ * @param multi_file_archive_path
+ * @param segment_ids
+ * @throws Propagates `update_offset`'s exceptions.
+ */
+auto write_archive_files(
+        FileWriter& archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        std::vector<std::string> const& segment_ids
+) -> void;
+
+auto get_segment_ids(segment_id_t next_segment_id) -> std::vector<std::string> {
```

Comment:
Thinking again, I feel that maybe we don't need this function?
Since we expect the segment_ids to be sequential, can't we just let get_file_infos and write_archive_files accepts an argument "last_segment_ids"?

It looks like segment_id = 0 is a special case, but there should be a way to write a logic the nicely handles it.